### DISCUSSION TITLE
handleSubmit/getValues is slow to trigger on forms with large data

### DISCUSSION THREAD

**Discussion by okbytes** [react-hook-form/react-hook-form#6560] [Q&A] [answered]
_I've been unable to create a complex enough form in an isolated scenario to reproduce this issue._

I have complex form data in this general shape:

```js
const formValues = {
  name: "Name",
  key1: { ...ComplexObject },
  key2: { ...ComplexObject },
  key3: { ...ComplexObject },
  configuration: { ...ComplexObject },
  fields: [
      { ...ComplexFieldObject },
      { ...ComplexFieldObject },
      { ...ComplexFieldObject } // ...and so on -> could be 300+ in this array
  ],
}
```

Calling `handleSubmit` or `getValues` results in a perceptible delay before the function executes its logic (and triggers an API call transforming the button into "Loading"). With form data that has `field.length < 150 or so`, the interaction appears instant and the "Loading" indicator appears as the user clicks.

What I am trying to understand is why these actions are taking so long and if I can do anything reasonable to speed up execution. Ideally, I'd be able to retain `handleSubmit` as the validation/error handling is helpful.

**Comment by bluebill1049** [comment]
which version are you running?

**Reply by okbytes** [reply]
v7.15.2

**Reply by bluebill1049** [reply]
This could be due to the validation attached to all your inputs. can you share a codesandbox?

**Reply by okbytes** [reply]
I am currently unable to get a working codesandbox that reproduces this issue. I have validation on `name` only, but I see the same delay when calling `getValues()` inside a function.

**Reply by bluebill1049** [reply]
getValues should definitely not getting any delay. there is 0 logic in that function.

**Reply by okbytes** [reply]
Could `useFormContext` be responsible for the slowness to trigger these functions? My form makes heavy use of `FormContext`.

**Reply by okbytes** [reply]
In doing some debugging in my codebase (locally), I discovered the delay when calling `handleSubmit`/`getValues` appears when the `fields.length >= 75`. Below that threshold, the execution feels instant in the UI.

**Reply by don-hover** [reply]
@bluebill1049 

I'm seeing a very similar issue, for a form that has a nested object data structure.

The form is in a list view, so the form data structure represents the "rows" in the list, but as a keyed object where the key is the `id` of the row's data item.  Like this:
```.ts
{
  100: { id:100, name: "Roof shingles", color: "Weathered Brown", colorId: "a234-599f", quantity: 4, sku: "968745" },
  102: { id:102, name: "Ridge shingles", color: "Weathered Brown", colorId: "a234-599c", quantity: 8, sku: "968746" },
  115: { id:102, name: "Flashing", color: "Brown", colorId: "b567-599f", quantity: 2, sku: "468755" },
  123: { id:123, name: "Nails", color: "1-inch", colorId: "f98c-599f", quantity: 1, sku: "1122745" },
}
```

It may seem redundant to keep the `id` as the key when it's also in the row data - but because this is an "editable table" UI, the rows are iterated over to render, and the input components must be registered with RHF during the iteration, and within the row components we often need to call `setValue()` or `getValue()` on a specific row's data, which would be challenging without a direct-access data structure like this.

What I'm seeing is that the data-entry UI of the text input components is compromised on the fields that are registered in a form with nested data like this.  When I type into a text input, it takes 1-2 seconds for the input to reflect the data that I've type.  These are uncontrolled components, and if I either un-register the input or change the registered `name` from a nested key to a flat/root key, the performance is as expected.

Can you provide any explanations of why this would be happening, or advice on ways to acheive the same thing without this UI performance problem?

**Comment by okbytes** [comment, marked as answer]
I have narrowed down the problem to re-renders in React. Mostly likely, it's how I've implemented react-hook-form in react-table.
